test/test_buf_error.c

```diff
@@ -0,0 +1,303 @@
+ * to finish at once.
+ * 3. test_scp - replicates an issue found using libnxz with scp.
+ */
+
+int test_0_avail_in() {
+
+        z_stream strm = {0};
+        Byte *compr = NULL, *uncompr = NULL;
+        unsigned int outlen = 4096;
+        int rc = 0, ret = 0;
+
+        if ((rc = inflateInit(&strm)) != 0){
+		printf("*** failed: inflateInt returned %d\n", rc);
+		return TEST_ERROR;
+	}
+
+        compr = (Byte*)calloc(1,1);
+	if ( compr == NULL ) {
+		printf("*** failed: failed to allocate memory");
+		return TEST_ERROR;
+	}
+
+        uncompr = (Byte*)calloc(outlen, sizeof(Byte));
+        if ( uncompr == NULL ) {
+		free(compr);
+		printf("*** failed: failed to allocate memory");
+		return TEST_ERROR;
+	}
+
+        strm.avail_out = outlen;
+        strm.next_out = uncompr;
+        strm.next_in = compr;
+
+	/* At this point avail_in is 0 */
+        rc = inflate(&strm, Z_PARTIAL_FLUSH);
+	if (rc != Z_BUF_ERROR) {
+		ret = TEST_ERROR;
+		printf("*** failed: inflate returned %d\n", rc);
+	}
+
+        if ((rc = inflateEnd(&strm)) != Z_OK) {
+		ret = TEST_ERROR;
+		printf("*** failed: inflateEnd returned %d\n", rc);
+	}
+
+	return ret;
+}
+
+int test_small_avail_out(){
+
+	z_stream strm = {0};
+	Byte *src, *compr, *uncompr;
+	int srclen = 128;
+	unsigned int comprlen = srclen*2;
+	unsigned int uncomprlen = 3;
+	int rc = 0, ret = 0;
+
+	generate_random_data(128);
+	src = ran_data;
+
+        compr = (Byte*)calloc(comprlen, sizeof(Byte));
+	if ( compr == NULL ) {
+		printf("*** failed: could not allocate memory");
+		return TEST_ERROR;
+	}
+
+        uncompr = (Byte*)calloc(uncomprlen, sizeof(Byte));
+        if ( uncompr == NULL ) {
+		free(compr);
+		printf("*** failed: could not allocate memory");
+		return TEST_ERROR;
+	}
+
+	if (_test_nx_deflate(src, srclen, compr, &comprlen, srclen, NULL) != 0) {
+		printf("*** failed: nx_deflate failed");
+		return TEST_ERROR;
+	}
+
+        if ((rc = inflateInit(&strm)) != 0){
+		printf("*** failed: inflateInt returned %d\n", rc);
+		return TEST_ERROR;
+	}
+
+        strm.avail_out = uncomprlen;
+        strm.next_out = uncompr;
+        strm.next_in = compr;
+	strm.avail_in = comprlen;
+
+	/* avail_out is not enough */
+        rc = inflate(&strm, Z_FINISH);
+	if (rc != Z_BUF_ERROR) {
+		ret = TEST_ERROR;
+		printf("*** failed: inflate returned %d\n", rc);
+	}
+
+        if ((rc = inflateEnd(&strm)) != Z_OK) {
+		ret = TEST_ERROR;
+		printf("*** failed: inflateEnd returned %d\n", rc);
+	}
+
+	return ret;
+}
+
+
+/*
+ * This test uses compressed data from scp to ensure the bug #74 is reproduced.
+ */
+int test_scp() {
+	Byte *uncompr;
+	int err;
+	unsigned int uncompr_len = 4096;
+	z_stream strm;
+
+	if ((uncompr = (Byte*)calloc((uInt)uncompr_len, 1)) == NULL)
+	{
+		printf("*** alloc buffer failed.\n");
+		return TEST_ERROR;
+
+	}
+
+
+	strm.zalloc = NULL;
+	strm.zfree = NULL;
+	strm.opaque = (voidpf) 0;
+
+	err = inflateInit(&strm);
+	if (err != Z_OK) {
```

Comment:
I see that throughout this file in some cases you use the construct 
```
if ((var = function()) != ret) { }
```
And others you use
```
var = function()
if (var != ret) { }
```

It's preferable to choose one, for consistency. 

I tend to prefer the second, even though more verbose it's clearer, but I'd be OK with any, as long as it's consistent.